tests/integration/relations/opensearch_provider/application-charm/src/charm.py

```diff
@@ -0,0 +1,239 @@
+        else:
+            logger.error("connection check to opensearch charm failed")
+            self.unit.status = BlockedStatus("No connection to opensearch charm")
+
+    def connection_check(self) -> bool:
+        """Simple connection check to see if backend exists and we can connect to it."""
+        relations = self.model.relations.get("first-database", []) + self.model.relations.get(
+            "second-database", []
+        )
+        if len(relations) == 0:
+            return False
+        for relation in relations:
+            if not self.smoke_check(relation.id):
+                return False
+        return True
+
+    def smoke_check(self, relation_id) -> bool:
+        try:
+            resp = self.relation_request(relation_id, "GET", "/")
+            return bool(resp)
+        except (OpenSearchHttpError, Exception) as e:
+            logger.error(e)
+            return False
+
+    # First database events observers.
+    def _on_first_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application."""
+        logging.info(f"first database credentials: {event.username} {event.password}")
+
+    def _on_first_database_endpoints_changed(self, event: DatabaseEndpointsChangedEvent) -> None:
+        """Event triggered when the read/write endpoints of the database change."""
+        logger.info(f"first database endpoints have been changed to: {event.endpoints}")
+
+    # Second database events observers.
+    def _on_second_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application."""
+        logger.info(f"second database credentials: {event.username} {event.password}")
+
+    def _on_second_database_endpoints_changed(self, event: DatabaseEndpointsChangedEvent) -> None:
+        """Event triggered when the read/write endpoints of the database change."""
+        logger.info(f"second database endpoints have been changed to: {event.endpoints}")
+
+    # ==============
+    #  Action hooks
+    # ==============
+    def _on_run_request_action(self, event: ActionEvent):
+        logger.info(event.params)
+        relation = self.first_database
+        relation_id = event.params["relation-id"]
+        databag = relation.fetch_relation_data()[relation_id]
+        method = event.params["method"]
+        endpoint = event.params["endpoint"]
+        payload = event.params.get("payload", None)
+        if payload:
+            payload = payload.replace("\\", "")
+
+        username = databag.get("username")
+        password = databag.get("password")
+        host = databag.get("endpoints").split(",")[0]
+        host_addr = host.split(":")[0]
+        port = host.split(":")[1]
+
+        logger.info(f"sending {method} request to {endpoint}")
+        try:
+            response = self.request(method, endpoint, port, username, password, host_addr, payload)
```

Comment:
I think here you need to cast the port into an int since it didn't happen before: `int(port)`